### PR TITLE
[WIP] ML check

### PR DESCRIPTION
**Pull request by szymonmaszke**
# ML check

# General things

- __Use a lot of markdown formatting__ (bullet lists, points, links, bold and so on). It looks
good and helps others to go on their own if they wish. Also the content is
rated higher in this case by the users (learned over StackOverflow). So spam formatting
whenever feasible
- Be shorter and less verbose (more bullet points)
- A little less comments (like there's no point in saying we import KNN from sklearn with a comment). With some other stuff we guide a little too much as well, keep in mind
- Connect math to it's representation more. For example: row in matrix -> example,
column in matrix -> feature
- When using bold: \*\*smth\*\* is deprecated, use \_\_smth\_\_

# [WIP] Missing

- reproducibility - importance of seeding & multiple runs for experimentation

# Module 1 - 1. Linear Regression (commit 1)

- Added full expression for linear regression with multiple features. Currently
added as Mathematical formula of model, might be put in other place.

### Firstly, how can we represent these examples numerically?

- Not sure about images but it will take too long to rewrite in TeX
- Notation is reversed; "We can stack these examples in columns to produce a **design matrix**, $X$";
usually (always?) examples are stacked as rows and columns are features

### 3. The criterion - How do we know how good our model is?

- Those are not synonyms, see [this answer](https://stats.stackexchange.com/a/179027/237147).
Not sure how to put it here, should probably just delete, left for now

### Curse of dimensionality

- Defo add a picture

### Challenges

- Definitely __do not__ put grid search here as it might be confusing. Separate
notebook for hyperparam search is needed (left for now).

# Module 1 - 2. Gradient Based Optimization (commit 2)

Soz for the wrong notebook modifications.
Not sure about model's name - `LinearHypothesis` seems like an odd choice

### The learning rate, $\alpha$

[TO-DO] Tbh I don't understand what we mean here. For example:

> We know that the gradient at a minima is zero, and at this point we want our parameters
to be moved with a step size of zero - so that they remain where they are, at the minima.

It doesn't matter what lr you set, it will be multiplied by zero gradient and we get zero

> If the step size were some constant value, then our model might need to adjust
its weights by some value smaller than this to reach a nearby minima.

LR is constant value, grad is changing but not sure what we are after here.

## OH NO WHAT ABOUT LOCAL OPTIMA?

Suddenly introduced local optima, nobody would think about it unless they already know it, but __minor__

> But most of the models that are useful in practice depend on many more parameters (neural networks can easily have millions).

Well, simple regression is still way more practical than DL

> And as the number of parameters increase, it becomes exponentially unlikely that any parameterisation is
at a minima, but is rather a saddle point, and so there is still an indication of how to improve.

Debatable, usually DL nets get to local minima, there are proofs for global minima
with SGD but it's not for this course. Also we should mention the goal __isn't__ to minimize
the function as is but to do the task on unseen data (this data may, and usually has, different minimum).
So Global Optimum should not be treated as a goal.

Also worth mentioning during the teaching that global minima does not mean the function
has zero value (heard that one), it might be large and never reach zero.

Something more about saddle points? Pic or smth

## Multivariate gradient descent

> A model with multiple parameters (such as a weight and a bias, or multiple weights)
would be optimised in the same way - we would just have more of these functions.

Cool, shouldn't we introduce logistic regression or any other model with more
than `1` param by this point?

## Implementing gradient descent from scratch

They won't be able to follow I'm afraid.

### Why not just pass a single datapoint to the model for each update?

- rewritten

### Why should we not pass the whole dataset through the model for each update?

This slide is split into list of elements. IMO more readable, check, verify, hit me up.

## Mini-batch gradient descent

> Most optimisation algorithms converge much faster if they are allowed to rapidly compute
approximate gradients rather than slowly compute exact gradients.

Not really. Stochasticity helps with regularization of the model. Also depends
on how much hardwarde power we got (that's why there are "mini" batches in millions of examples).
Something about RAM limitation as well?

### Code for it

Talk about seeding, random numbers & some stochasticity?

> We can see here that stochastic and mini-batch training perform inference (making predictions) >3x faster and make updates 2.5x faster than
full batch training. For larger datasets or inputs, these differences would be exaggerated even further.
This is because the model has to pass the whole dataset forward to make a prediction and then compute the average gradient from every
single one of those values to compute the weight and bias updates.

Something about cache locality, levels of memory __AND__ vectorization like SIMD
under the hood? And that we need and equilibrium for good performance. Just a few basic words

# Module 2- 0. K nearest neighbours

## Whole code

Students should "think" in numpy so we should force them to do so. No other way around
it. Loops are evil. This is a bad habit they will have to leave anyway. Otherwise
looks fine (maybe update TeX, but __minor__)

# Module 2- 1. K-Means [TBD]

- [TBD] __Major code rewrite__ - KMeans should be sklearn-compatible like other `class` oriented algos
- Missing initialization schemes for K-Means (K-Means++), multiple iterations, metrics how to measure whether the cluster is good (we can give some metrics here and some over at DBScan)
- All in all this lesson could be vastly expanded
- Challenge is too much of a headache with 3D visualization in matplotlib.

In general this one IMO needs a major rewrite

# Module 2- 2. DBSCAN [TBD]

- Same as above
- Calculate distances with `numpy`, not with loops

### CODE REVIEW

**Comment by life-efficient** [comment]
- [ ] Can we find pictures online for where suggested?
- [ ] K-means code was some of the first stuff we ever taught lol - hence how it's written. Let's refactor and update the challenge to something you deem more appropriate
- [ ] Would love for you to add in words about hardware specs cacheing, RAM etc. Not something i know a lot about
- [ ] deffo replace loops with numpy where appropriate lol

**Comment by szymonmaszke** [comment]
@life-efficient Other things:

## Change ML Foundations split into modules

Currently the structure is as follows:

- Machine Learning Concepts
	- Overview
	- Linear Regression
	- Gradient Based Optimization
- Classification & Regression
	- Multivariate regression and feature normalisation
	- Bias, variance and generalisation
	- Hyperparameters and validation sets
	- Regularisation basics
	- MLE
	- Multiclass classification
- Clustering
	- K-Means
	- DBSCAN
- CARTs
	- Classification Trees
	- Regression Trees
- SVMs
	- SVM

Caveats I got with this structure:
- too long/short modules; Classification & Regression have 6 notebooks, SVMs have one
- strange module split; why SVMs have their own module, why do CARTs while Classification & Regression is in one place?

Structure that seems more reasonable would be something along those lines (__to add in bold__, _moved in italic_). Also some lessons and modules are renamed to better reflect the content:

- Machine Learning in general
	- Overview
	- Linear Regression
	- Gradient Based Optimization
	- _Bias, variance, overfitting, underfitting_
	- _Regularisation_
	- _Validation and Test sets_ (__split in two with below__)
	- _Hyperparameters and Grid Search_
	- MLE
- Supervised learning
	- Multivariate regression and feature normalisation
	- Multiclass classification
	- SVMs
	- Classification Trees (regression trees should be removed IMO)
- Ensembles & boosting
	- Bagging & Random Forests
	- Boosting & Adaboost
	- Gradient Boosting
	- XGBoost
- Unsupervised learning
	- K-Means
	- DBSCAN
	- __PCA__
	- __T-SNE__
